074-additional-volumes-support.md

```diff
@@ -0,0 +1,95 @@
+# Support of additional volumes in pod
+
+We propose to enhance the Strimzi Kafka Operator by adding support for specifying additional volumes in the CRD. This will involve modifying the Kafka CRDs to include an `additionalVolumes` field, which will allow users to define extra volumes and volume mounts.
+
+The specific usecase of ours, is that we want to set custom log4j logging for audit logs, to ensure that audit logs are *not* present in stdout, but can be picked up by a sidecar which uses OpenTelemtry or logstash to forward to the audit logs to a secure location.
+
+## Current situation
+
+The operator handles many configurations and operational aspects but lacks support for specifying additional volumes in the Custom Resource Definition (CRD). This limitation can restrict users from accessing log files or integrating with external storage solutions directly.
+
+## Motivation
+
+Adding support for additional volumes in the CRD will allow users to:
+
+ - Access log files for enhanced monitoring and debugging.
+ - Use additional configurations or secret management tools that require mounting extra volumes.
+
+Improve flexibility in managing Kafka deployments by allowing users to attach custom storage solutions based on their specific requirements.
+
+## Proposal
+
+### Considerations
+
+- CRD Changes: Update the Kafka CRDs to include an `additionalVolumes` field.
+- Validation: Ensure that the additional volumes are validated correctly and do not conflict with existing volume definitions such as `/var/lib/kafka/` and `/tmp`.

```

Comment:
I think this needs more detail as there might be more to it than just duplicate mount paths.
* Duplicate volume names?
* Duplicate or conflicting paths as already mentioned?
* What if the same volume is mounted multiple times? Does it have any implications?

Reply:
This is quite an interesting discussion. Looking at other operators and how the kubeapi works, I would propose not to add too much validation here. If the platform engineer defines the custom resource in a invalid way, then kubernetes will do validation and throw errors when reconciling. E.g telling the user that the volume with same name already exists etc.

I do not think we need to reproduce this validation logic from kubeapi into this operator. However, I do see an argument for validating cases such as adding a volume that already conflicts with the *existing* volumes that the operator adds such as `/tmp`. I hope the distinction makes sense.

Reply:
I think validating the name is fairly simple and easy to do. So that should be done by the operator. You are right that Kubernetes will validate these things. But it will do so only once the original Pod is deleted which weakens the availability. So I think it needs to be evaluated case by case what does or doesn't make sense -> for names and mount path I think the validation is reasonably simple and should be done.

Reply:
I still think this should go into more details what will be validated, what will be the possible check for the volume path etc.

Reply:
As discussed in the community call, we will limit the path to `/mnt` to ensure backwards compatibility further